ql/src/java/org/apache/hadoop/hive/ql/exec/repl/util/SnapshotUtils.java

```diff
@@ -0,0 +1,415 @@
+   * @param snapshotPrefix prefix used by the snapshot
+   * @param conf Hive Configuration.
+   * @param snapshotCount the counter to store number of deleted snapshots.
+   * @param isLoad If this is called for clearing up snapshots at target cluster, In case of load it renames the
+   *               snapshot file as well.
+   * @throws IOException
+   * @throws SemanticException
+   */
+  public static void cleanupSnapshots(Path dumpRoot, String snapshotPrefix, HiveConf conf,
+      ReplSnapshotCount snapshotCount, boolean isLoad) throws IOException, SemanticException {
+    DistributedFileSystem dfs = (DistributedFileSystem) dumpRoot.getFileSystem(conf);
+    if (dfs.exists(new Path(dumpRoot, EximUtil.FILE_LIST_EXTERNAL_SNAPSHOT_OLD))) {
+      FileList snapOld = createTableFileList(dumpRoot, EximUtil.FILE_LIST_EXTERNAL_SNAPSHOT_OLD, conf);
+      FileList snapNew = createTableFileList(dumpRoot, EximUtil.FILE_LIST_EXTERNAL_SNAPSHOT_CURRENT, conf);
+      ArrayList<String> oldPaths = SnapshotUtils.getListFromFileList(snapOld);
+      ArrayList<String> newPaths = SnapshotUtils.getListFromFileList(snapNew);
+      ArrayList<String> diffList = SnapshotUtils.getDiffList(newPaths, oldPaths, conf, isLoad);
+      dfs = isLoad ? (DistributedFileSystem) getExternalTableBaseDir(conf).getFileSystem(conf) : dfs;
+      SnapshotUtils.cleanUpSnapshots(dfs, diffList, snapshotPrefix, snapshotCount);
+    }
+    if (isLoad) {
+      try {
+        dfs.delete((new Path(dumpRoot, EximUtil.FILE_LIST_EXTERNAL_SNAPSHOT_OLD)), true);
+      } catch (FileNotFoundException fnf) {
+        // ignore
+        LOG.warn("Failed to clean up snapshot " + EximUtil.FILE_LIST_EXTERNAL_SNAPSHOT_OLD, fnf);
+      }
+      try {
+        dfs.rename(new Path(dumpRoot, EximUtil.FILE_LIST_EXTERNAL_SNAPSHOT_CURRENT),
+            new Path(dumpRoot, EximUtil.FILE_LIST_EXTERNAL_SNAPSHOT_OLD), Options.Rename.OVERWRITE);
+      } catch (FileNotFoundException fnf) {
+        // ignore
+        LOG.warn("Failed to clean up snapshot " + EximUtil.FILE_LIST_EXTERNAL_SNAPSHOT_CURRENT, fnf);
+      }
+    }
+  }
+
+  public static boolean shouldRdiff(Path p, HiveConf conf, String snapshot) {
+    try {
+      DistributedFileSystem dfs = getDFS(p, conf);
+      // check if the target got modified
+      int index =
+          dfs.getClient().getSnapshotDiffReportListing(p.toUri().getPath(), snapshot, "", DFSUtilClient.EMPTY_BYTES, -1)
+              .getLastIndex();
+
+      return index > -1;
+    } catch (Exception e) {
+      return false;
+    }
+  }
+
+  public static void deleteReplRelatedSnapshots(FileSystem fs, Path path) {
+    try {
+      FileStatus[] listing = fs.listStatus(new Path(path, ".snapshot"));
+      for (FileStatus elem : listing) {
+        if (elem.getPath().getName().contains(OLD_SNAPSHOT) || elem.getPath().getName().contains(NEW_SNAPSHOT)) {
+          deleteSnapshotSafe((DistributedFileSystem) fs, path, elem.getPath().getName());
```

Comment:
can we leak snapshots. we dont compare if delete was successful or not

Reply:
Changed